system/security/cryptohelper/ske.cpp

```diff
@@ -0,0 +1,161 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+/*
+ * Module dealing with openssl symmetric key encryption/decription.
+ * For now just AES.
+ */
+
+#if defined(_USE_OPENSSL) && !defined(_WIN32)
+
+#include "jliball.hpp"
+#include <openssl/pem.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+
+#include "cryptocommon.hpp"
+#include "pke.hpp"
+#include "ske.hpp"
+
+namespace cryptohelper
+{
+
+size32_t aesKeyEncrypt(MemoryBuffer &out, size32_t inSz, const void *inBytes, const char key[aesKeySize], const char iv[aesBlockSize])
+{
+    OwnedEVPCipherCtx ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        throw makeEVPException(0, "Failed EVP_CIPHER_CTX_new");
+
+    /* Initialise the encryption operation. IMPORTANT - ensure you use a key
+     * and IV size appropriate for your cipher
+     * In this example we are using 256 bit AES (i.e. a 256 bit key). The
+     * IV size for *most* modes is the same as the block size. For AES this
+     * is 128 bits
+     * */
+    if (1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), nullptr, (const unsigned char *)key, (const unsigned char *)iv))
+        throw makeEVPException(0, "Failed EVP_EncryptInit_ex");
+
+    /* Provide the message to be encrypted, and obtain the encrypted output.
+     * EVP_EncryptUpdate can be called multiple times if necessary
+     */
+
+    const size32_t cipherBlockSz = 128;
+    size32_t outMaxSz = inSz + cipherBlockSz/8;
+    size32_t startSz = out.length();
+    byte *outPtr = (byte *)out.reserveTruncate(outMaxSz);
+    int outSz;
+    if (1 != EVP_EncryptUpdate(ctx, (unsigned char *)outPtr, &outSz, (unsigned char *)inBytes, inSz))
+        throw makeEVPException(0, "Failed EVP_EncryptUpdate");
+    int ciphertext_len = outSz;
+
+    /* Finalise the encryption. Further ciphertext bytes may be written at
+     * this stage.
+     */
+    if (1 != EVP_EncryptFinal_ex(ctx, outPtr + outSz, &outSz))
+        throw makeEVPException(0, "Failed EVP_EncryptFinal_ex");
+    ciphertext_len += outSz;
+    out.setLength(startSz+ciphertext_len); // truncate length of 'out' to final size
+    return (size32_t)ciphertext_len;
+}
+
+size32_t aesKeyDecrypt(MemoryBuffer &out, size32_t inSz, const void *inBytes, const char *key, const char *iv)
+{
+    OwnedEVPCipherCtx ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        throw makeEVPException(0, "Failed EVP_CIPHER_CTX_new");
+
+    const size32_t cipherBlockSz = 128;
+    // man page says(!) - "should have sufficient room for (inl + cipher_block_size) bytes unless the cipher block size is 1 in which case inl bytes is sufficient
+    size32_t outMaxSz = ((inSz==1?8:inSz) + cipherBlockSz/8);
```

Comment:
This code does not match the previous comment - it is checking insz rather than the cipher block size.  There should be some unit tests that check encrypting and decrypting 0, 1, 15 (128/8-1), 16 and 17 bytes.

Reply:
hm, I was misreading the man page/comment, will change.

Will add some unit tests.